lib/experimental/html/class-wp-html-character-reference-transcoder.php

```diff
@@ -0,0 +1,1149 @@
+					}
+
+					// Ambiguous ampersand is context-sensitive.
+					switch ( $context ) {
+						case 'attribute':
+							$buffer .= substr( $input, $next, $at - $next );
+							continue 3;
+
+						case 'data':
+							$buffer .= substr( $group, $sub_at, $sub_length );
+							continue 3;
+					}
+				}
+			}
+
+			/*
+			 * Whether by failing to find a group or failing to find a name,
+			 * we have failed to match a character reference name, so we can
+			 * continue processing as if this is plain text and leave the
+			 * invalid character reference name in place.
+			 */
+			$buffer .= substr( $input, $next, $at );
+		}
+
+		if ( 0 === $at ) {
+			return $input;
+		}
+
+		if ( $at < $end ) {
+			$buffer .= substr( $input, $at );
+		}
+
+		return $buffer;
+	}
+
+	/**
+	 * When numeric character references indicate certain code points
+	 * within the C1 control characters, these substitutions are used.
+	 *
+	 * @see https://html.spec.whatwg.org/#numeric-character-reference-end-state
+	 *
+	 * @var string[]
+	 */
+	static $c1_replacements = array(
+		0x80 => "€",
+		0x82 => "‚",
+		0x83 => "ƒ",
+		0x84 => "„",
+		0x85 => "…",
+		0x86 => "†",
+		0x87 => "‡",
+		0x88 => "ˆ",
+		0x89 => "‰",
+		0x8A => "Š",
+		0x8B => "‹",
+		0x8C => "Œ",
+		0x8E => "Ž",
+		0x91 => "‘",
+		0x92 => "’",
+		0x93 => "“",
+		0x94 => "”",
+		0x95 => "•",
+		0x96 => "–",
+		0x97 => "—",
+		0x98 => "˜",
+		0x99 => "™",
+		0x9A => "š",
+		0x9B => "›",
+		0x9C => "œ",
+		0x9E => "ž",
+		0x9F => "Ÿ",
+	);
+
+	/**
+	 * Autogenerated table of named character references.
+	 *
+	 * Character reference names are grouped by their first two characters,
+	 * then each group is string encoding the names and substitutions in
+	 * that group.
+	 *  - The first byte represents the length of the following name.
+	 *  - The bytes pertaining to the name follow the first byte.
+	 *  - The next byte represents the length of the following substitution for the previous name.
+	 *  - The bytes pertaining to the substitution follow this byte.
+	 *  - The end of the group string indicates that there are no more names in the group.
+	 *
+	 * The format for this lookup table is the result of an attempt to minimize memory
+	 * overhead while maintaining fast lookup. Since every named character reference is
+	 * at least three letters long (shortest is one of `&GT`, `&LT`, `&gt`, `&lt`), we
+	 * take the first two non-`&` letters as a group, leaving an average of 3.5 names
+	 * per group with a median of 2 names per group.
+	 *
+	 * The small size of each group should preserve memory locality of the group names
+	 * while maintaining quick testing of each candidate in the group since we know the
+	 * entire group and all its data will occupy consecutive bytes in memory.
+	 *
+	 * @see https://html.spec.whatwg.org/#named-character-references
+	 * @see https://html.spec.whatwg.org/entities.json
+	 *
+	 * @var string[] named character reference information
+	 */
+	public static $lookup_table = array(
```

Comment:
Is there any way to initialize this on the first use and avoid allocating the memory until the first use? E.g. implementing something like...

```php
public static get_lookup_table() {
    if ( ! self::$lookup_table ) {
        self::$lookup_table = /* ... */;
    }
    return self::$lookup_table;
}
```

Or perhaps PHP JIT already delays the memory allocation until the first use of `$lookup table`? I couldn't be sure.

Reply:
as static string literals I don't think this is possible, or that it will save any memory, because those strings already have to be in memory for it to lazily-initialize them.

this is down to a negligible amount IMO and constrasts a few MBs from clearer constructs, such as the array you mention before, such as the array introduced in d663e70f8205f06e2437ec5280ea2220edcc0018